Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -2610,6 +2680,113 @@ namespace Babylon
+
+                for (const auto& id : squeezeStarts)
+                {
+                    auto inputSourceFound = m_idToInputSource.find(id);
+                    if (inputSourceFound != m_idToInputSource.end())
+                    {
+                        auto inputSourceVal = inputSourceFound->second.Value();
+                        Napi::Object inputSourceEvent = GenerateXRInputSourceEvent(inputSourceVal, env);
+
+                        for (const auto& [name, callback] : m_eventNamesAndCallbacks)
+                        {
+                            if (name == JS_EVENT_NAME_SQUEEZE_START)
+                            {
+                                callback.Call({inputSourceEvent});
+                            }
+                        }
+
+                        m_activeSqueezes.insert(id);
+                    }
+                }
+
+                for (const auto& id : selectEnds)
+                {
+                    auto inputSourceFound = m_idToInputSource.find(id);
+                    if (inputSourceFound != m_idToInputSource.end())
+                    {
+                        auto inputSourceVal = inputSourceFound->second.Value();
+                        Napi::Object inputSourceEvent = GenerateXRInputSourceEvent(inputSourceVal, env);
+
+                        for (const auto& [name, callback] : m_eventNamesAndCallbacks)
+                        {
+                            if (name == JS_EVENT_NAME_SELECT_END)
+                            {
+                                callback.Call({inputSourceEvent});
+                            }
+
+                            if (name == JS_EVENT_NAME_SELECT)
+                            {
+                                callback.Call({inputSourceEvent});
+                            }
+                        }
+
+                        m_activeSelects.erase(id);
+                    }
+                }
+
+                for (const auto& id : squeezeEnds)
+                {
+                    auto inputSourceFound = m_idToInputSource.find(id);
+                    if (inputSourceFound != m_idToInputSource.end())
+                    {
+                        auto inputSourceVal = inputSourceFound->second.Value();
+                        Napi::Object inputSourceEvent = GenerateXRInputSourceEvent(inputSourceVal, env);
+
+                        for (const auto& [name, callback] : m_eventNamesAndCallbacks)
+                        {
+                            if (name == JS_EVENT_NAME_SQUEEZE_END)
+                            {
+                                callback.Call({inputSourceEvent});
+                            }
+
+                            if (name == JS_EVENT_NAME_SQUEEZE)
+                            {
+                                callback.Call({inputSourceEvent});
+                            }
+                        }
+
+                        m_activeSqueezes.erase(id);
+                    }
+                }
```

Comment:
These are all basically the same actions, just with different event names and source collections. Can this be extracted out into a separate function or at least a common lambda?